### DISCUSSION TITLE
Let TreeView support Virtualize

### DISCUSSION THREAD

**Discussion by stefandhondt** [microsoft/fluentui-blazor#2804] [Ideas]
When working with large datasets, expanding a treeview item which contains thousands of children will severely impact performance, even when lazy loading children.
It would be preferred to have the Virtualize-property available and have it work the same as it does with FluentDataGrid where only the visible children will be created.

**Comment by vnbaaij** [comment]
I'm not sure if that can work with a TreeView. The premise of Virtualize is that each item needs to have the same height. As a tree view by nature is a recursive structure, the heights will vary. Also, we would need to have a kind of 'viewport' where parts of the tree are not shown (otherwise Virtualize is not going to help) and we don't do that by default.
We need to test it to make sure but it is not high on the priority list. Help is appreciated.

**Reply by stefandhondt** [reply]
I know about the ItemSize-property of DatGrid and my thinking to overcome this was to insert/remove rows at the appropriate location in the items collection on expand/collapse. This way, all items of the treeview still have the same size.
Since I already have a need for this on my current project, I'm in the process of creating a custom component that will hopefully look/act like a treeview but is basically a datagrid, so I get the virtualization for free.

**Reply by vnbaaij** [reply]
> I know about the ItemSize-property of DatGrid and my thinking to overcome this was to insert/remove rows at the appropriate location in the items collection on expand/collapse. This way, all items of the treeview still have the same size.

Don't think that will work because the items are nested. 

> Since I already have a need for this on my current project, I'm in the process of creating a custom component that will hopefully look/act like a treeview but is basically a datagrid, so I get the virtualization for free.

Please post something here if you get it all working

**Comment by stefandhondt** [comment]
![2804](https://github.com/user-attachments/assets/eb08645a-b9ad-44f8-bd96-db04418b7d4c)
[2804.zip](https://github.com/user-attachments/files/17375969/2804.zip)

**Reply by stefandhondt** [reply]
@vnbaaij : Thoughts on my implementation?

**Reply by vnbaaij** [reply]
I've not been able to review yet.
